### PR TITLE
[new release] tls, tls-mirage, tls-lwt, tls-eio and tls-async (0.17.0)

### PR DESCRIPTION
**Pull request by hannesm**
Transport Layer Security purely in OCaml

- Project page: <a href="https://github.com/mirleft/ocaml-tls">https://github.com/mirleft/ocaml-tls</a>
- Documentation: <a href="https://mirleft.github.io/ocaml-tls/doc">https://mirleft.github.io/ocaml-tls/doc</a>

##### CHANGES:

* tls-async: remove ocaml < 5.0.0 constraint (mirleft/ocaml-tls#471 mirleft/ocaml-tls#474 @hannesm)
* remove dependency on ppx, especially ppx_cstruct and ppx_sexp_conv across
  the libraries (mirleft/ocaml-tls#473 @hannesm, discussion in mirleft/ocaml-tls#472)

### CODE REVIEW

**Comment by hannesm** [comment]
the reverse dependencies with an upper bound in the second commit:
- https://github.com/Psi-Prod/Mehari/pull/64
- https://github.com/mirage/ocaml-conduit/pull/421

**Comment by hannesm** [comment]
CI failures are unrelated

File: packages/tls-eio/tls-eio.0.17.0/opam [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+bug-reports:  "https://github.com/mirleft/ocaml-tls/issues"
+doc:          "https://mirleft.github.io/ocaml-tls/doc"
+authors:      ["Thomas Leonard"]
+maintainer:   ["Hannes Mehnert <hannes@mehnert.org>" "David Kaloper <david@numm.org>"]
+license:      "BSD-2-Clause"
+
+build: [
+  ["dune" "subst"] {dev}
+  ["dune" "build" "-p" name "-j" jobs]
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
```
Comment (kit-ty-kate) [inline packages/tls-eio/tls-eio.0.17.0/opam:13]:
```suggestion
  ["dune" "runtest" "-p" name "-j" jobs] {with-test & os != "macos"}
```
```
#=== ERROR while compiling tls-eio.0.17.0 =====================================#
# context              2.1.4 | macos/x86_64 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/mirleft/ocaml-tls/releases/download/v0.17.0/tls-0.17.0.tbz)
# path                 ~/.opam/5.0.0/.opam-switch/build/tls-eio.0.17.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune runtest -p tls-eio -j 11
# exit-code            1
# env-file             ~/.opam/log/tls-eio-33607-41d193.env
# output-file          ~/.opam/log/tls-eio-33607-41d193.out
### output ###
# (cd _build/default && /Users/mac1000/.opam/5.0.0/bin/ocamlopt.opt -w -40 -g -o eio/tests/fuzz.exe /Users/mac1000/.opam/5.0.0/lib/cmdliner/cmdliner.cmxa /Users/mac1000/.opam/5.0.0/lib/ocplib-endian/ocplib_endian.cmxa /Users/mac1000/.opam/5.0.0/lib/ocaml/unix/unix.cmxa /Users/mac1000/.opam/5.0.0/lib/afl-persistent/afl-persistent.cmxa /Users/mac1000/.opam/5.0.0/lib/ocaml/str/str.cmxa /Users/mac1000/.opam/5.0.0/lib/crowbar/crowbar.cmxa /Users/mac1000/.opam/5.0.0/lib/cstruct/cstruct.cmxa -I /Users/mac1000/.opam/5.0.0/lib/cstruct /Users/mac1000/.opam/5.0.0/lib/logs/logs.cmxa /Users/mac1000/.opam/5.0.0/lib/eqaf/eqaf.cmxa /Users/mac1000/.opam/5.0.0/lib/eqaf/bigstring/eqaf_bigstring.cmxa /Users/mac1000/.opam/5.0.0/lib/eqaf/cstruct/eqaf_cstruct.cmxa /Users/mac1000/.opam/5.0.0/lib/mirage-crypto/mirage_crypto.cmxa -I /Users/mac1000/.opam/5.0.0/lib/mirage-crypto /Users/mac1000/.opam/5.0.0/lib/hkdf/hkdf.cmxa /Users/mac1000/.opam/5.0.0/lib/mirage-crypto-rng/mirage_crypto_rng.cmxa /Users/mac1000/.opam/5.0.0/lib/zarith/zarith.cmxa -I /Users/mac1000/.opam/5.0.0/lib/zarith /Users/mac1000/.opam/5.0.0/lib/sexplib0/sexplib0.cmxa /Users/mac1000/.opam/5.0.0/lib/mirage-crypto-pk/mirage_crypto_pk.cmxa /Users/mac1000/.opam/5.0.0/lib/ptime/ptime.cmxa /Users/mac1000/.opam/5.0.0/lib/asn1-combinators/asn1_combinators.cmxa /Users/mac1000/.opam/5.0.0/lib/fmt/fmt.cmxa /Users/mac1000/.opam/5.0.0/lib/gmap/gmap.cmxa /Users/mac1000/.opam/5.0.0/lib/domain-name/domain_name.cmxa /Users/mac1000/.opam/5.0.0/lib/base64/base64.cmxa /Users/mac1000/.opam/5.0.0/lib/mirage-crypto-ec/mirage_crypto_ec.cmxa -I /Users/mac1000/.opam/5.0.0/lib/mirage-crypto-ec /Users/mac1000/.opam/5.0.0/lib/pbkdf/pbkdf.cmxa /Users/mac1000/.opam/5.0.0/lib/macaddr/macaddr.cmxa /Users/mac1000/.opam/5.0.0/lib/ipaddr/ipaddr.cmxa /Users/mac1000/.opam/5.0.0/lib/x509/x509.cmxa /Users/mac1000/.opam/5.0.0/lib/tls/tls.cmxa /Users/mac1000/.opam/5.0.0/lib/hmap/hmap.cmxa /Users/mac1000/.opam/5.0.0/lib/lwt-dllist/lwt_dllist.cmxa /Users/mac1000/.opam/5.0.0/lib/optint/optint.cmxa /Users/mac1000/.opam/5.0.0/lib/eio/core/eio__core.cmxa /Users/mac1000/.opam/5.0.0/lib/bigstringaf/bigstringaf.cmxa -I /Users/mac1000/.opam/5.0.0/lib/bigstringaf /Users/mac1000/.opam/5.0.0/lib/mtime/mtime.cmxa /Users/mac1000/.opam/5.0.0/lib/eio/eio.cmxa /Users/mac1000/.opam/5.0.0/lib/ptime/clock/os/ptime_clock.cmxa -I /Users/mac1000/.opam/5.0.0/lib/ptime/clock/os eio/tls_eio.cmxa /Users/mac1000/.opam/5.0.0/lib/psq/psq.cmxa /Users/mac1000/.opam/5.0.0/lib/eio/utils/eio_utils.cmxa /Users/mac1000/.opam/5.0.0/lib/eio/mock/eio_mock.cmxa /Users/mac1000/.opam/5.0.0/lib/logs/logs_fmt.cmxa eio/tests/.fuzz.eobjs/native/dune__exe.cmx eio/tests/.fuzz.eobjs/native/dune__exe__Mock_socket.cmx eio/tests/.fuzz.eobjs/native/dune__exe__Test_rng.cmx eio/tests/.fuzz.eobjs/native/dune__exe__Fuzz.cmx)
# ld: warning: directory not found for option '-L/opt/local/lib'
# File "eio/tests/tls_eio.md", line 1, characters 0-0:
# /usr/local/bin/git --no-pager diff --no-index --color=always -u _build/default/eio/tests/tls_eio.md _build/default/eio/tests/.mdx/tls_eio.md.corrected
# diff --git a/_build/default/eio/tests/tls_eio.md b/_build/default/eio/tests/.mdx/tls_eio.md.corrected
# index 285acea..ac5343a 100644
# --- a/_build/default/eio/tests/tls_eio.md
# +++ b/_build/default/eio/tests/.mdx/tls_eio.md.corrected
# @@ -102,11 +102,5 @@ let serve_ssl ~config server_s callback =
#         test_client ~net ("127.0.0.1", "4433")
#      )
#    ;;
# -+server -> start @ tcp:127.0.0.1:4433
# -+server -> connect
# -+handler accepted
# -+handler + GET / HTTP/1.1
# -+client <- GET
# -+client done.
# -- : unit = ()
# +Exception: Eio.Io Eio_luv.Luv_error(EPERM) (* operation not permitted *)
#  ```
# (cd _build/default/eio/tests && ./fuzz.exe --repeat 200)
# random ops: PASS
# 
```

**Comment by kit-ty-kate** [comment]
`tls-eio` uses the network and thus can’t be tested on macOS. Could you return the fix upstream?

We’ll merge this asap

**Comment by hannesm** [comment]
> `tls-eio` uses the network and thus can’t be tested on macOS. Could you return the fix upstream?

sure, https://github.com/mirleft/ocaml-tls/commit/fdd2f0ef3982c456f54e057c6ea9b232649ed92d

**Comment by kit-ty-kate** [comment]
Thanks!

**Pull request merged**